### PR TITLE
Rework of reboot plugin, and lots of tests

### PR DESCRIPTION
**Pull request by jonasfj**
Filed against `worker2` branch so that it's not so big :)

Half the patch is declarative tests, don't worry too much about those...

### CODE REVIEW

File: plugins/reboot/reboot.go [hunk 1]
```diff
@@ -1,161 +1,174 @@
...
+		p.Plugin.rebooted.Do(func() {
+			p.Monitor.Infof("Stopping worker after task as instructed by task.payload.reboot = '%s'", p.RebootAction)
+			p.Plugin.Worker.StopGracefully()
+		})
 	}
+	return nil
+}
 
+func (p *taskPlugin) Exception(runtime.ExceptionReason) error {
+	if p.RebootAction == rebootOnFailure || p.RebootAction == rebootOnException {
```
Comment (walac) [inline plugins/reboot/reboot.go]: What about `rebootAlways`?

File: plugins/reboot/reboot_test.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+			"allowTaskReboots": true
+		}`,
+		Payload: `{
+			"delay": 10,
+			"function": "false",
+			"argument": "",
+			"reboot": "on-exception"
+		}`,
+	}.Test()
+}
```
Comment (walac) [inline plugins/reboot/reboot_test.go]: Add test case for when payload is `always` and an exception happens.

File: plugins/reboot/reboot.go (hunk 1 above)
Comment (jonasfj) [inline plugins/reboot/reboot.go]: Great catch!!! :) thanks...

File: plugins/reboot/reboot_test.go (hunk 2 above)
Comment (jonasfj) [inline plugins/reboot/reboot_test.go]: Fixed, thanks for pointing the bug! :)

**Comment by jonasfj** [comment]
ignore test failure... it's just linter timeouts...

**Pull request merged**